File: src/smt_sa/fifo.py
from collections import deque


class FIFO:
    def __init__(self, name):
        self.name = name

        self._arch_fifo = deque()
        self._phy_fifo = deque()

        self._pushed = False
        self._popped = False
        self._last_pushed = None
        self._arch_fifo_size = 0
        self._phy_fifo_size = 0

    def peek(self, pos=0):
        if self._arch_fifo.__len__() == 0:
            return None
        else:
            return self._arch_fifo[pos]

    def push(self, x):
        if self._pushed:
            raise ValueError("Can't push to the queue more than once per cycle")

        self._phy_fifo.append(x)
        self._pushed = True
        self._last_pushed = x
        self._phy_fifo_size += 1

    def pop(self):
        if self._popped:
            raise RuntimeError("Can't pop from the queue more than once per cycle")

        if self._phy_fifo_size > 0:
            self._phy_fifo.popleft()
            self._phy_fifo_size -= 1

        self._popped = True
        return self.peek()

    def size(self):
        return self._arch_fifo_size

    def cycle(self):
        if self._pushed:
            self._arch_fifo.append(self._last_pushed)
            self._pushed = False
            self._last_pushed = None

        if self._popped:
            self._arch_fifo.popleft()
            self._popped = False

        self._arch_fifo_size = self._phy_fifo_size

File: Baseline-SMT-SA/baseline_smt_sa/src/fifo.cpp
#ifndef FIFO_CPP
#define FIFO_CPP

#include <iostream>
#include <string>
#include <queue>
#include <cassert>

using namespace std;

template <typename T>
class fifo {
private:
    string _name;

    T _last_pushed;
    bool _pushed;
    bool _popped;
public:
    queue<T> _arch_fifo;
    queue<T> _phy_fifo;
    
    fifo (string name);
    T peek(bool& is_empty);
    void push(T x);
    T pop(bool& is_empty);
    uint32_t size();
    void cycle();
};

template <typename T>
fifo<T>::fifo (string name) : _name(name), _pushed(false), _popped(false) {}

template <typename T>
T fifo<T>::peek(bool& is_empty) {
    if (_arch_fifo.size() == 0) {
        is_empty = true;
        // TODO: should probably return something
    }
    else {
        is_empty = false;
        return _arch_fifo.front();
    }
}

template <typename T>
void fifo<T>::push(T x) {
    assert(_pushed == false);

    _phy_fifo.push(x);
    _pushed = true;
    _last_pushed = x;
}

template <typename T>
T fifo<T>::pop(bool& is_empty) {
    assert(_popped == false);

    if (_phy_fifo.size() > 0)
        _phy_fifo.pop();

    _popped = true;
    return peek(is_empty);
}

template <typename T>
uint32_t fifo<T>::size() {
    return _arch_fifo.size();
}

template <typename T>
void fifo<T>::cycle() {
    if (_pushed) {
        _arch_fifo.push(_last_pushed);
        _pushed = false;
    }

    if (_popped) {
        _arch_fifo.pop();
        _popped = false;
    }

    assert(_arch_fifo.size() == _phy_fifo.size());
}

#endif